Userland/Libraries/LibWeb/HTML/FormAssociatedElement.cpp

```diff
@@ -152,61 +165,248 @@ void FormAssociatedElement::reset_form_owner()
+    }
+
+    // 2. Set the selection range with the value of this element's selectionStart attribute, the
+    //    given value, and the value of this element's selectionDirection attribute.
+    return set_selection_range(m_selection_start, value, selection_direction());
+}
+
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#selection-direction
+Optional<String> FormAssociatedElement::selection_direction() const
+{
+    switch (m_selection_direction) {
+    case SelectionDirection::Forward:
+        return "forward"_string;
+    case SelectionDirection::Backward:
+        return "backward"_string;
+    case SelectionDirection::None:
+        return "none"_string;
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#set-the-selection-direction
+void FormAssociatedElement::set_selection_direction(Optional<String> direction)
 {
-    // 1. If this element is an input element, and selectionEnd does not apply to this element, throw an
-    //    "InvalidStateError" DOMException.
-    // NOTE: This is done by HTMLInputElement before calling this function
+    // To set the selection direction of an element to a given direction, update the element's
+    // selection direction to the given direction, unless the direction is "none" and the
+    // platform does not support that direction; in that case, update the element's selection
+    // direction to "forward".
+    m_selection_direction = string_to_selection_direction(direction);
+}
+
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#dom-textarea/input-setselectionrange
+WebIDL::ExceptionOr<void> FormAssociatedElement::set_selection_range(Optional<WebIDL::UnsignedLong> start, Optional<WebIDL::UnsignedLong> end, Optional<String> direction)
+{
+    // 1. If this element is an input element, and setSelectionRange() does not apply to this
+    //    element, throw an "InvalidStateError" DOMException.
+    auto& html_element = form_associated_element_to_html_element();
+    if (is<HTMLInputElement>(html_element) && !static_cast<HTMLInputElement&>(html_element).selection_or_range_applies())
+        return WebIDL::InvalidStateError::create(html_element.realm(), "setSelectionRange does not apply to this input type"_fly_string);
 
-    // FIXME: 2. Set the selection range with the value of this element's selectionStart attribute, the given value, and the
-    //           value of this element's selectionDirection attribute.
+    // 2. Set the selection range with start, end, and direction.
+    set_the_selection_range(start, end, string_to_selection_direction(direction));
     return {};
 }
 
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#set-the-selection-range
+void FormAssociatedElement::set_the_selection_range(Optional<WebIDL::UnsignedLong> start, Optional<WebIDL::UnsignedLong> end, SelectionDirection direction)
```

Comment:
Normally what we do on optional parameters (`direction`) that have a default fallback is to just default the parameter at the declaration site:
```c++
void set_the_selection_range(..., SelectionDirection direction = SearchDirection::None);
```

Doing that here seems like it'd align us closer to the spec, since we could then invoke this AO directly, rather than going through the `setSelectionRange` binding. And it'd avoid checking the element type twice.

And since `selection_direction` returns the attribute string, we'd also add a getter for internal use to return its enum value. We do this for HTMLInputElement's type state, for example.

Reply:
There are two different specs, one is for `setSelectionRange` and calls into the second spec for `set the selection range`. A spec note on UI interaction calls for selection updates to also call into `set the selection range`, not `setSelectionRange`. That's why they're now separate; they more closely follow the spec.

Reply:
Oh, I didn't mean to combine them into one. I just meant to make this one more easily callable from IDL methods. Because right now, the IDL methods for `select`, `setSelectionStart` and `setSelectionEnd` invoke the `setSelectionRange` variant, whereas their spec steps say to invoke the `set the selection range` variant.

So, if:
1. The `set the selection range` impl has a default `direction` parameter of none
2. We have a `selection_direction_state` (or whatever) that returns the enum value directly, rather than as a string

Then the callsite from `setSelectionStart` (and similarly for `setSelectionEnd`) looks like:
```c++
// 4. Set the selection range with the given value, end, and the value of this element's
//    selectionDirection attribute.
return set_the_selection_range(value, end, selection_direction_state());
```

And the callsite from `select` looks like:
```c++
// 2. Set the selection range with 0 and infinity.
return set_the_selection_range(0, NumericLimits<WebIDL::UnsignedLong>::max());
```

Reply:
Right, this allows for a better fit to the spec at the expense of a somewhat clunky `selection_direction_state()` but that seems fine. Thanks!